#########################################################################
# ⭐⭐⭐  Set parameters carefully for satisfied performance!   ⭐⭐⭐ #
#########################################################################

mass        : 0.979 # kg 
gra         : 9.81 
pose_solver : 1     # 0:From ZhepeiWang (drag & less singular) 1:From ZhepeiWang, 2:From rotor-drag    
ctrl_freq_max    : 100.0
use_bodyrate_ctrl: true
max_manual_vel   : 1.0
max_angle: 80  # Attitude angle limit in degree. A negative value means no limit.
low_voltage: 13.2 # 4S battery
odom_freq: 100
ctrl_mode: 2      # DFBC 1 MPC 2 JPCM 3

rc_reverse: # *
    roll: false
    pitch: true
    yaw: true
    throttle: true

auto_takeoff_land:
    enable: true
    enable_auto_arm: true
    no_RC: false
    takeoff_height: 1.0 # m
    takeoff_land_speed: 0.20 # m/s

thrust_model: # The model that maps thrust signal thr_bodyrate_u(0~1) to real thrust force F(Unit:N): F=K1*Voltage^K2*(K3*thr_bodyrate_u^2+(1-K3)*thr_bodyrate_u). 
    print_value: false # display the value of “thr_scale_compensate” or “hover_percentage” during thrust model estimating.
    accurate_thrust_model: false  # This can always enabled if don't require accurate control performance :-)
    # accurate thrust mapping parameters
    K1: 0.7583 # Needs precise calibration!
    K2: 1.6942 # Needs precise calibration!
    K3: 0.6786 # Needs precise calibration! K3 equals THR_MDL_FAC in https://docs.px4.io/master/en/config_mc/pid_tuning_guide_multicopter.html.
    # approximate thrust mapping parameters
    hover_percentage: 0.23  # Thrust percentage in Stabilize/Arco mode # *
    thrust_upper_bound: 0.60
    thrust_lower_bound: 0.00
    
gain: 
    # Thrust PID
    Kp0: 8.0
    Kp1: 8.0 
    Kp2: 10.0
    Kv0: 4.0
    Kv1: 4.0
    Kv2: 6.0
    # Kp0: 4.0
    # Kp1: 4.0 
    # Kp2: 5.0
    # Kv0: 2.0
    # Kv1: 2.0
    # Kv2: 3.0
    # ↓↓↓ No use now --
    Kvi0: 0.0
    Kvi1: 0.0
    Kvi2: 0.0
    Kvd0: 0.0
    Kvd1: 0.0
    Kvd2: 0.0
    # ↓↓↓ Only used in rate control mode.
    KAngR: 6.0
    KAngP: 6.0
    KAngY: 4.0
    # max error
    PErrMax: 0.5
    VErrMax: 1.0

rotor_drag:  
    # x: 0.24  # The reduced acceleration on each axis caused by rotor drag. Unit:(m*s^-2)/(m*s^-1).
    # y: 0.22  # Same as above
    # z: 0.31  # Same as above
    # k_thrust_horz: 0.005 #0.007 # Set to 0 recommended... --
    x: -10.0  # The reduced acceleration on each axis caused by rotor drag. Unit:(m*s^-2)/(m*s^-1). k
    y: -2.23   # Same as above D
    z: -0.33 # Same as above D1
    k_thrust_horz: 0.00 #0.007 # Set to 0 recommended... --

msg_timeout:
    odom: 0.5
    rc:   0.5
    cmd:  0.1
    imu:  0.5
    bat:  0.5

Factor_graph:
    LOG_NAME: TGYRO

    PRI_VICON_POS_COV: 0.001
    PRI_VICON_VEL_COV: 0.001

    CONTROL_P_COV_X:   0.15
    CONTROL_PF_COV_X:  0.05
    CONTROL_P_COV_Y:   0.15
    CONTROL_PF_COV_Y:  0.05
    CONTROL_P_COV_Z:   0.20
    CONTROL_PF_COV_Z:  0.05

    CONTROL_R3_COV:    0.03
    CONTROL_R2_COV:    0.3
    CONTROL_R1_COV:    0.3

    CONTROL_V_COV:     2.3
    DYNAMIC_P_COV:     0.0001
    DYNAMIC_R_COV:     0.0001
    DYNAMIC_V_COV:     0.0001
    
    INPUT_JERK_T:      0.2
    INPUT_JERK_M:      0.1
    INPUT_JERK_M3:     0.01
    # ↓↓↓ limit Thrust and angular speed
    CLF_HIGH:          27
    CLF_LOW:           5
    G_CLF_HIGH:        3 
    G_CLF_LOW:        -3
    CLF_THR:           1
    G_CLF_THR:         0.1
    CLF_ALPHA:         1

    OPT_LENS_TRAJ: 20
    WINDOW_SIZE:   5

    POS_MEAS_MEAN:      0.000
    POS_MEAS_COV:       0.010 # 0.05
    VEL_MEAS_COV:       0.050
    ROT_MEAS_COV:       0.010
    
    PRIOR_POS_MEAS_COV: 0.01 # 0.05
    PRIOR_VEL_MEAS_COV: 0.05
    PRIOR_ROT_MEAS_COV: 0.10

    # IMU Params
    acc_sigma_x:        0.1 # 0.3
    acc_bias_imu_x:     0.00004
    
    acc_sigma_y:        0.1
    acc_bias_imu_y:     0.001

    acc_sigma_z:        0.1
    acc_bias_imu_z:     0.001

    gyro_sigma_x:       0.01 # 0.02
    gyro_bias_sigma_x:  0.000018

    gyro_sigma_y:       0.03
    gyro_bias_sigma_y:  0.001

    gyro_sigma_z:       0.01
    gyro_bias_sigma_z:  0.0001

    # acc_sigma:        0.01
    # gyro_sigma:       0.001
    # acc_bias_imu:     0.001
    # gyro_bias_sigma:  0.0001

    prior_acc_sigma:    0.01
    prior_gyro_sigma:   0.01

    opt_gravity_rot:    false
    prior_gravity_roll:  0.00
    prior_gravity_pitch: 0.00

    use_vel:            false
    use_rot:            true

